ui/src/app/applications/components/application-status-panel/application-status-panel.tsx

```diff
@@ -47,79 +47,112 @@ const sectionHeader = (info: SectionInfo, hasMultipleSources: boolean, onClick?:
 };
 
 export const ApplicationStatusPanel = ({application, showDiff, showOperation, showConditions, showExtension, showMetadataInfo}: Props) => {
-    const today = new Date();
-
+    let cntByCategory;
+    let hasMultipleSources;
+    let source;
+    let appOperationState: models.OperationState;
     let daysSinceLastSynchronized = 0;
-    const history = application.status.history || [];
-    if (history.length > 0) {
-        const deployDate = new Date(history[history.length - 1].deployedAt);
-        daysSinceLastSynchronized = Math.round(Math.abs((today.getTime() - deployDate.getTime()) / (24 * 60 * 60 * 1000)));
-    }
-    const cntByCategory = (application.status.conditions || []).reduce(
-        (map, next) => map.set(getConditionCategory(next), (map.get(getConditionCategory(next)) || 0) + 1),
-        new Map<string, number>()
-    );
-    const appOperationState = getAppOperationState(application);
-    if (application.metadata.deletionTimestamp && !appOperationState) {
-        showOperation = null;
+
+    if (isApp) {
```

Comment:
Just a feeling because I am not sure what complexity it would add on the frontend, but the code is harder to read with all the `if (isApp) {}` conditions. I feel it's in these situations that the abstraction does not fit and it should be 2 different components, with shared functions.

Reply:
Thank you for your feedback!
I am also unhappy with the isApp(). Can you give me an example (for the backend, I am also new to React so trying to draw parallels)? Something like a factory design pattern on the backend?

Reply:
I was thinking something like that. Basically, everywhere isApp is used, then you should have `AbstractApp` define a method and each abstraction (App and AppSet) implement it.

For the object construction, It seems to be in the services files (applications-service.ts) where you would need to perform creation. See the `parseAppFields` method. Right now, is seems to just cast it in the abstract interface, but I think building an abstract type with a factory would be the right way to build the objects.

```ts
interface AppInterface {
    something: string;
}

abstract class AbstractApp implements AppInterface {
    something: string;

    constructor(something: string) {
        this.something = something;
    }

    display(): string {
        return "This is a common method"
    }

    abstract doLogic(): string;
}

class Application extends AbstractApp {
    name: string;
    other: string;

    constructor(name: string, other: string) {
        super("app")
        this.name = name;
        this.other = other;
    }
    doLogic(): string {
        return `Application: ${this.name} with ${this.other}`;
    }
}

class ApplicationSet extends AbstractApp {
    name: string;

    constructor(name: string) {
        super("appset")
        this.name = name
    }
    doLogic(): string {
        return `AppSet: ${this.name}`;
    }
}


// Before
var apps:AppInterface[] = [
    {something: "app", name: "one", other: "magic"} as AppInterface,
    {something: "appset", name: "two"} as AppInterface
] 
apps.forEach(function (app) {
    console.log(isApp(app) ? `Application: ${(app as Application).name} with ${(app as Application).other}` : `AppSet: ${(app as ApplicationSet).name}`)
});

function isApp(app: AppInterface) {
    return app.something == "app"
}


// After 
var apps2:AbstractApp[] = [
    appFactory({something: "app", name: "one", other: "magic"}),
    appFactory({something: "appset", name: "two"}),
] 
apps2.forEach(function (app) {
    console.log(app.doLogic())
});

function appFactory(app: any): AbstractApp {
    if (app?.something == "appset") {
        return new ApplicationSet(app.name)
    } 
    return new Application(app.name, app.other)
}
```

Reply:
Yeah, there used to be a lot more of these conditional statements (in Regina's branch) and the goal is to move away from this.  The approach was to do it in stages (if else will minimize possibility of regressions) and then refactor further from there.  The code has always been using class components (versus function components) so we're going to continue to use it to our advantage.